### PR TITLE
Preparations for a CAD-based arithmetic solver

### PR DESCRIPTION
**Pull request by nafur**
Based on #4679, this PR contains further preparations for a CAD-based arithmetic solver that extends the current `NonlinearExtension`.
In detail, it provides:
- a `Constraints` class that manages all (polynomial) constraints visible to the cad solver,
- a collection of methods used for cad projections,
- a `VariableOrdering` class that provides different variable ordering heuristics tailored to cad,
- an extension to the `NlModel` class, allowing for witness terms,
- further conversion methods, in particular between `Node` and `poly::Polynomial`, `poly::Value` and `RealAlgebraicNumber`.

### CODE REVIEW

File: src/theory/arith/nl/nl_model.cpp [hunk 1]
```diff
@@ -362,6 +363,11 @@ bool NlModel::addCheckModelSubstitution(TNode v, TNode s)
       return false;
     }
   }
+  auto itw = d_check_model_witnesses.find(v);
+  if (itw != d_check_model_witnesses.end())
+  {
+    // TODO(Gereon): Check whether s satisfies itw->second.
```
Comment (ajreynol) [inline src/theory/arith/nl/nl_model.cpp:369]: We usually associate TODO with issues/wishues, for example see https://github.com/CVC4/CVC4/blob/master/src/theory/arith/nl/nl_model.h#L178.

File: src/theory/arith/nl/nl_model.cpp [hunk 2]
```diff
@@ -1087,13 +1115,23 @@ bool NlModel::simpleCheckModelMsum(const std::map<Node, Node>& msum, bool pol)
...
-          Trace("nl-ext-cms")
-              << "  failed due to unknown bound for " << vc << std::endl;
-          // should either assign a model bound or eliminate the variable
-          // via substitution
-          Assert(false);
-          return false;
+          auto wit = d_check_model_witnesses.find(vc);
+          if (wit != d_check_model_witnesses.end())
+          {
+            // TODO(Gereon): Should do something here.
```
Comment (ajreynol) [inline src/theory/arith/nl/nl_model.cpp:1121]: Ditto.

File: src/theory/arith/nl/nl_model.h [hunk 3]
```diff
@@ -203,7 +211,8 @@ class NlModel
    */
   void getModelValueRepair(
       std::map<Node, Node>& arithModel,
-      std::map<Node, std::pair<Node, Node>>& approximations);
+      std::map<Node, std::pair<Node, Node>>& approximations,
```
Comment (ajreynol) [inline src/theory/arith/nl/nl_model.h:214]: Add documentation to this method for the new argument.

File: src/theory/arith/nl/nl_model.cpp (hunk 2 above)
Comment (nafur) [reply src/theory/arith/nl/nl_model.cpp:1121]: Not worrying and not doing anything sounds good to me.

File: src/theory/arith/nl/nl_model.cpp (hunk 1 above)
Comment (nafur) [reply src/theory/arith/nl/nl_model.cpp]: As per offline discussion, this case should not happen and can thus be excluded (by assertion).

File: src/theory/arith/nl/cad/constraints.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implements a container for CAD constraints.
+ **
+ ** Implements a container for CAD constraints.
+ **/
+
+#ifndef CVC4__THEORY__NLARITH__CAD__CONSTRAINTS_H
```
Comment (ajreynol) [inline src/theory/arith/nl/cad/constraints.h:23]: This guard should be `CVC4__THEORY__ARITH__NL__CAD__CONSTRAINTS_H` (namespaces caps with two underscores separating), similarly format the others.

Comment (nafur) [reply src/theory/arith/nl/cad/constraints.h]: I've renamed the guards.

**Comment by aniemetz** [comment]
(In the header files, is what I'm saying. And for short comments in the header files (that fit within 80 chars) we put the final `*/` on the same line.)

File: src/theory/arith/nl/poly_conversion.cpp [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+  Trace("poly::conversion")
+      << "Converting " << p << " over " << var << std::endl;
+
+  std::vector<poly::Integer> coeffs = coefficients(p);
+
+  auto* nm = NodeManager::currentNM();
+
+  Node res = nm->mkConst(Rational(0));
+  Node monomial = nm->mkConst(Rational(1));
+  for (std::size_t i = 0; i < coeffs.size(); ++i)
```
Comment (aniemetz) [inline src/theory/arith/nl/poly_conversion.cpp:83]: Ditto

File: src/theory/arith/nl/nl_model.h [hunk 6]
```diff
@@ -128,6 +128,14 @@ class NlModel
    * bounds.
    */
   bool addCheckModelBound(TNode v, TNode l, TNode u);
+  /** add check model witness
```
[the remark this answers is not in the document]
Comment (nafur) [reply src/theory/arith/nl/nl_model.h]: Indeed. I changed it for all comments in this file.

File: src/theory/arith/nl/cad/variable_ordering.h [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+#include "util/poly_util.h"
+
+namespace CVC4 {
+namespace theory {
+namespace arith {
+namespace nl {
+namespace cad {
+
+/** Variable orderings for real variables in the context of CAD. */
+enum class VariableOrderingStrategy
```
[the remark this answers is not in the document]
Comment (nafur) [reply src/theory/arith/nl/cad/variable_ordering.h:42]: Fixed.

File: src/theory/arith/nl/cad/projections.cpp [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+  for (const auto& p : polys)
+  {
+    for (const auto& coeff : coefficients(p))
+    {
+      add_polynomial(res, coeff);
+    }
+    add_polynomial(res, discriminant(p));
+  }
+  for (std::size_t i = 0; i < polys.size(); ++i)
```
[the remark this answers is not in the document]
Comment (nafur) [reply src/theory/arith/nl/cad/projections.cpp]: I hope I've fixed all relevant loops...

File: src/theory/arith/nl/cad/constraints.h [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+   */
+  VariableMapper mVarMapper;
+
+  void sort_constraints();
+
+ public:
+  VariableMapper& var_mapper() { return mVarMapper; }
+
+  /**
+   * Add a constraing (represented by a polynomial and a sign condition) to the
```
Comment (aniemetz) [inline src/theory/arith/nl/cad/constraints.h:71]:
```suggestion
   * Add a constraint (represented by a polynomial and a sign condition) to the
```

**Pull request merged**